[FEATURE] Enables dynamic linking with MKL and compiler based OpenMP

akarbown opened the pull request:
OneMKL 2021.3 fixed linking OpenMP while using SDL and MKL_THREADING_LAYER set to GNU.

## Description ##
OneMKL 2021.3 fixes the issue described [here](https://github.com/apache/incubator-mxnet/issues/18255#issuecomment-817830091). Thus, it enables linking with MKL dynamic libraries without having multiple OneMPs in a single process. It is possible due to linking MxNET with oneMKL Single Dynamic Library (SDL) and then setting the appropriate threading layer at run time in a function mkl_threading_layer() (or through environment variable MKL_THREADING_LAYER).

Connected with: [#19610], [#18255] and [#17794].

### Changes ###	

1. Add oneMKL 2021.3 to ubuntu docker images.
2. Enable MKL SDL (MKL_USE_SINGLE_DYNAMIC_LIBRARY) as the default linking when MKL version is grower than 2021.2 and static linking is turned off. (Bug no: MKLD-11109, [OneMKL release notes](https://software.intel.com/content/www/us/en/develop/articles/intel-oneapi-math-kernel-library-onemkl-bug-fixes.html)) .
4. Otherwise, MKL static libraries are taken into account and used to build MxNET library.
5. Add support of the new oneMKL file structure in the FindBLAS.cmake file (fix comes from the cmake 3.20: [#6210](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6210) ).


## Comments ##
Does using oneMKL 2021.3 as the recommended one should be mentioned in the documentation?

leezu commented on src/initialize.cc:
```diff
@@ -93,6 +95,13 @@ LibraryInitializer::LibraryInitializer()
+  #if defined( __INTEL_LLVM_COMPILER)
+    mkl_set_threading_layer(MKL_THREADING_INTEL);
+  #else
+    mkl_set_threading_layer(MKL_THREADING_GNU);
```
Does this work with Windows? Intel developer's reference states "for GNU threading on Linux* operating system only" https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/support-functions/single-dynamic-library-control/mkl-set-threading-layer.html

akarbown replied:
Good point! I forgot about Windows! I'll exclude it in a moment.

akarbown commented:
@mxnet-bot run ci[sanity]

akarbown commented:
@mxnet-bot run ci [all]

akarbown commented:
@mxnet-bot run ci[all]

akarbown commented:
@leezu, could you help me with rerunning 'sanity' check in this PR? I've checked it locally and I don't see any issues. I suppose that this sanity check failed because of the timeout (SIGTERM). Is it possible?

leezu commented:
@josephevans can we extend the max time for sanity? This PR triggers rebuild of the Docker used for Sanity, and apparently thus timeouts

akarbown commented:
@mxnet-bot run ci [unix-cpu]

akarbown commented:
@mxnet-bot run ci[website]

akarbown commented:
@mxnet-bot run ci[clang, miscellaneous, unix-cpu, unix-gpu]

akarbown commented:
@mxnet-bot run ci[unix-cpu]

akarbown commented:
@mxnet-bot run ci[unix-gpu, miscellaneous]

akarbown commented:
@mxnet-bot run ci[miscellaneous]

akarbown commented:
@mxnet-bot run ci[miscellaneous]

akarbown commented:
@mxnet-bot run ci[unix-cpu, unix-gpu]

akarbown commented:
@mxnet-bot run ci[unix-cpu]

akarbown commented:
@mxnet-bot run ci[unix-cpu, unix-cpu]

akarbown commented:
@mxnet-bot run ci[centos-cpu]

akarbown commented:
@mxnet-bot run ci[unix-cpu]

leezu commented:
@mxnet-bot run ci[centos-gpu, unix-cpu]

mozga-intel commented:
Hi @akarbown It looks really good. Thanks! I have a question. 

Basically, there are two ways: either we'll use MKL_THREADING_GNU or MKL_THREADING_INTEL. Could you please tell me whether there are differences in performance results, or, there is no impact on that?

akarbown commented:
> Hi @akarbown It looks really good. Thanks! I have a question.
> 
> Basically, there are two ways: either we'll use MKL_THREADING_GNU or MKL_THREADING_INTEL. Could you please tell me whether there are differences in performance results, or, there is no impact on that?

@mozga-intel, this boils down to comparison between ICX and GCC compilers. The whole point of this PR is mainly to enable using dynamic linking of MKL libraries without hang or different OpenMP symbol conflicts. Thus performance of this 2 compilers is not really an issue here. Unit tests for both compilers run more or less in the same time (on my local machine). I've observed 10 more tests failures for icx than for gcc (unittest_ubuntu_python3_cpu), but it seems to be because of the fact that it wasn't tested before and there were problem with compiling MxNET with icc at all.

akarbown commented:
@mxnet-bot run ci[centos-gpu]

bgawrych commented on src/initialize.cc:
```diff
@@ -93,6 +95,24 @@ LibraryInitializer::LibraryInitializer()
     mp_cv_num_threads_(dmlc::GetEnv("MXNET_MP_OPENCV_NUM_THREADS", 0)) {
   dmlc::InitLogging("mxnet");
+
+#if !(defined(_WIN32) || defined(_WIN64) || defined(__WINDOWS__))
```
maybe it should be moved to a function like install_pthread_atfork_handlers - it probably will be more descriptive on what's happening here

mozga-intel commented:
> > Hi @akarbown It looks really good. Thanks! I have a question.
> > Basically, there are two ways: either we'll use MKL_THREADING_GNU or MKL_THREADING_INTEL. Could you please tell me whether there are differences in performance results, or, there is no impact on that?
> 
> @mozga-intel, this boils down to comparison between ICX and GCC compilers. The whole point of this PR is mainly to enable using dynamic linking of MKL libraries without hang or different OpenMP symbol conflicts. Thus performance of this 2 compilers is not really an issue here. Unit tests for both compilers run more or less in the same time (on my local machine). I've observed 10 more tests failures for icx than for gcc (unittest_ubuntu_python3_cpu), but it seems to be because of the fact that it wasn't tested before and there were problem with compiling MxNET with icc at all.

Thanks, Anna! I have one more question: it might not exactly precisely aims at an issue like this. Please have a look at the scenario like this: If I have huge matrices then I can use one of those functionalities: either sparse or dense BLAS. 
If I try to use the IIL64 interface for huge matrices on a 32-bit architecture, will I get either an error message or everything is okay? (...) 32-bits architecture doesn't support the ILP64 interface, and LP64 works only for a small tensor. Could be possible that it works for dense matrices but for a sparse I will get an error? Or, maybe the matrices will be divided into small matrices [?]

akarbown reviewed (commented):
Done.

akarbown commented:
@mxnet-bot run ci[unix-cpu]

akarbown commented:
> > > Hi @akarbown It looks really good. Thanks! I have a question.
> > > Basically, there are two ways: either we'll use MKL_THREADING_GNU or MKL_THREADING_INTEL. Could you please tell me whether there are differences in performance results, or, there is no impact on that?
> > 
> > 
> > @mozga-intel, this boils down to comparison between ICX and GCC compilers. The whole point of this PR is mainly to enable using dynamic linking of MKL libraries without hang or different OpenMP symbol conflicts. Thus performance of this 2 compilers is not really an issue here. Unit tests for both compilers run more or less in the same time (on my local machine). I've observed 10 more tests failures for icx than for gcc (unittest_ubuntu_python3_cpu), but it seems to be because of the fact that it wasn't tested before and there were problem with compiling MxNET with icc at all.
> 
> Thanks, Anna! I have one more question: it might not exactly aim at an issue like this. Please have a look at the scenario like this: If I have huge matrices then I can use one of both modes: either sparse or dense BLAS. What would I get, If I tried to call the IIL64 interface for huge matrices on a 32-bit architecture? Will I get an error message or will everything be okay? (...) 32-bits architecture doesn't support the ILP64 interface, and LP64 works only for a small tensor.

@mozga-intel Thanks for the questions!
MxNET MKL ILP64 (USE_INT64_TENSOR_SIZE) is a cmake dependent option and depends on the architecture. Thus, if you do not force that option to 'ON' it for sure on 32-bit architecture is 'OFF' (please look at CMakeLists.txt l.313). FindBLAS.cmake for 32-bit architecture sets BLA_VENDOR to 'Intel10_32' and depending on the SDL option links the following libraries:
1. SDL=1: ```Found BLAS: /opt/intel/oneapi/lib/ia32/libmkl_rt.so;-lpthread;-lm;-ldl```
2. SDL=0: ```Found BLAS: -Wl,--start-group;/opt/intel/oneapi/lib/ia32/libmkl_intel.so;/opt/intel/oneapi/lib/ia32/libmkl_intel_thread.so;/opt/intel/oneapi/lib/ia32/libmkl_core.so;-Wl,--end-group;/opt/intel/oneapi/lib/ia32/libiomp5.so;-lpthread;-lm;-ldl```

Those libraries are linked independently of the following options: BLA_VENDOR or USE_INT64_TENSOR_SIZE. The error won't appear as the architecture and the libraries are chosen correctly (as far as I've checked). 
I've asked you on the IM for the command line you want me to run and it would be really nice to get it. However, I don't see 32-bit testing and building scenario in the MxNET but I might have overlooked it - could you please provide me also which configuration: OS/docker you thought about?

I've checked that MxNET has some building issues on the 32-bit architecture thus I do not think that spending more time on that issue is in the scope of this PR.

akarbown commented:
@mxnet-bot run ci[unix-cpu]

akarbown commented on cmake/ChooseBlas.cmake:
```diff
@@ -155,14 +164,19 @@ elseif(BLAS STREQUAL "MKL" OR BLAS STREQUAL "mkl")
     endif()
   endif()
   # Setting up BLAS_mkl_MKLROOT for non-Ubuntu 20.04 OSes
-  find_path(BLAS_mkl_MKLROOT mkl PATHS $ENV{MKLROOT} ${INTEL_HOME_ROOT} ${INTEL_OPT_ROOT})
```
Fixed as suggested.

akarbown commented:
@mxnet-bot run ci[centos-gpu, unix-cpu]

bgawrych reviewed (approved):
LGTM

akarbown commented:
@mxnet-bot run ci[centos-gpu]

leezu commented:
@akarbown what's the status of this PR? cc @barry-jin who is tracking the 2.0 beta release

akarbown commented:
> @akarbown what's the status of this PR? cc @barry-jin who is tracking the 2.0 beta release

IMHO, it's done! @szha suggested to check if it works on Mac. As I haven't got any MacOS close at hand I'm trying to do it as github action.

akarbown commented:
@mxnet-bot run ci[centos-gpu]

akarbown commented:
@mxnet-bot run ci[unix-cpu]

akarbown commented:
@mxnet-bot run ci[unix-cpu]

akarbown commented:
@szha, @leezu, @barry-jin - can I ask you for the review of that PR?

I've added new github action pipeline to test and verify this change with oneMKL on MacOS. It revealed some things that were needed to fix for that OS i.e:
- remove using `` -Wl, --start-group/ -Wl, --end-group`` while linking static MKL libraries in FindBLAS.cmake (issue mentioned here: https://gitlab.kitware.com/cmake/cmake/-/issues/20548);
- set proper threading layer at runtime. According to the [documentation](https://software.intel.com/content/dam/develop/external/us/en/documents/onemkl-developerguide-mac.pdf) it need to be set mkl_set_threading_layer(MKL_THREADING_INTEL);
- added github action (using existing [os_x_staticbuild.yml](https://github.com/apache/incubator-mxnet/blob/master/.github/workflows/os_x_staticbuild.yml)) that is building OneMKL with static MKL libraries (to be consistent with already existing scripts) and run the same tests that are for os_x_staticbuild.yml plus MKL tests;
- fixed hangs that appeared while running those tests were the result of the numpy linking/using OpenBLAS instead of MKL BLAS and as a consequence it was linking libgomp which resulted in the hang (two OpenMP runtimes in one process). Recompiling it (done in the numpy_mkl.sh file) resolved the issue;
- excluded ``test_bf16_operator`` tests for that action pipeline as CI MacOS seems to not support avx512;
- tested locally MxNET linked with static, dynamic and SDL (Single Dynamic Library) on MacOS and all the tests (from the os_x_staticbuild.yml + MKL tests) seems to pass without any hang.

Now it seems the change seems to be tested and checked for MacOS and with MKL BLAS. 
Do you think that leaving that new github action for MKL on MacOS make sense? If so, can it look as it is or change it somehow?

**Remark**: I see that windows-gpu fails, but it's rather not connected with that change but maybe with the VS 2019 version 16.11 Release? As I see that for v16.8.1 (MSVC 19.28.29333.0) it [passed](https://jenkins.mxnet-ci.amazon-ml.com/blue/rest/organizations/jenkins/pipelines/mxnet-validation/pipelines/windows-gpu/branches/PR-20474/runs/12/nodes/40/steps/84/log/?start=0) without any issues, while for v16.11.4 (MSVC 19.29.30136.0) it [fails](https://jenkins.mxnet-ci.amazon-ml.com/blue/rest/organizations/jenkins/pipelines/mxnet-validation/pipelines/windows-gpu/branches/PR-20474/runs/13/nodes/40/steps/84/log/?start=0). But I'm not 100% sure.

barry-jin commented:
> Now it seems the change seems to be tested and checked for MacOS and with MKL BLAS.
> Do you think that leaving that new github action for MKL on MacOS make sense? If so, can it look as it is or change it somehow?

Yes, we need to keep this new github action as it will cover MKL related build and tests on MacOS.

szha commented:
@mxnet-bot run ci [windows-gpu]

akarbown commented:
@mxnet-bot run ci[unix-cpu, unix-gpu, centos-gpu]

akarbown commented:
@mxnet-bot run ci[centos-gpu]

leezu commented:
Thank you @akarbown!
